plan/attributes.go

```diff
@@ -2,34 +2,198 @@ package plan
+
+	for i, reqAttrMap := range reqAttrsSlice {
+		for _, reqAttr := range reqAttrMap {
+			ppred := node.Predecessors()[i].(*PhysicalPlanNode)
+			haveAttr, n := getOutputAttributeWithNode(node.Predecessors()[i].(*PhysicalPlanNode), reqAttr.Key())
+			if haveAttr == nil {
+				return errors.Newf(codes.Internal,
+					"attribute %q, required by %q, is missing from predecessor %q",
+					reqAttr.Key(), node.ID(), n.ID(),
+				)
+			}
+
+			if !reqAttr.SatisfiedBy(haveAttr) {
+				return errors.Newf(codes.Internal,
+					"node %q requires attribute %v, which is not satisfied by predecessor %q, "+
+						"which has attribute %v",
+					node.ID(), reqAttr, ppred.ID(), haveAttr,
+				)
+			}
+		}
+	}
 
-type ParallelRunAttribute struct {
-	Factor int
+	return nil
 }
 
-// If a node produces parallel data, then all successors must require parallel
-// data, otherwise there will be a plan error.
-func (ParallelRunAttribute) SuccessorsMustRequire() bool {
-	return true
+// GetOutputAttribute will return the attribute with the given key
+// provided by the given plan node, traversing backwards through predecessors
+// as needed for attributes that may pass through. E.g.,
+//   sort |> filter
+// The "filter" node will still provide the collation attribute, even though it's
+// the "sort" node that actually does the collating.
+func GetOutputAttribute(node *PhysicalPlanNode, attrKey string) PhysicalAttr {
+	attr, _ := getOutputAttributeWithNode(node, attrKey)
+	return attr
 }
 
-type ParallelMergeAttribute struct {
-	Factor int
+func getOutputAttributeWithNode(node *PhysicalPlanNode, attrKey string) (PhysicalAttr, *PhysicalPlanNode) {
+	if attr, ok := node.outputAttrs()[attrKey]; ok {
+		return attr, nil
+	}
+
+	if node.passesThroughAttr(attrKey) && len(node.Predecessors()) == 1 {
+		// TODO(cwolff): consider what it means for nodes with multiple predecessors
+		//   (e.g. join or union) to pass on attributes.
+		return getOutputAttributeWithNode(node.Predecessors()[0].(*PhysicalPlanNode), attrKey)
+	}
+
+	return nil, node
 }
 
-func (ParallelMergeAttribute) SuccessorsMustRequire() bool {
-	return false
+// CheckSuccessorsMustRequire will return an error if the node has an output attribute
+// that must be required by a successor, and nil otherwise.. E.g., the parallel-run
+// attribute is like this in that it must be required by a merge node.
+// This function will walk forward through successors to find the requiring node.
+func CheckSuccessorsMustRequire(node *PhysicalPlanNode) error {
+	for _, attr := range node.outputAttrs() {
+		if !attr.SuccessorsMustRequire() {
+			continue
+		}
+
+		if len(node.Successors()) == 0 {
+			return &flux.Error{
+				Code: codes.Internal,
+				Msg: fmt.Sprintf("node %q provides attribute %v that must be required but has no "+
+					"successors to require it", node.ID(), attr.Key()),
+			}
+		}
+		for _, succ := range node.Successors() {
+			if reqd, n := requiredBySuccessor(attr, node, succ.(*PhysicalPlanNode)); !reqd {
```

Comment:
When I've seen nesting of conditionals, esp. in a loop,  I have been trying to find ways to make them go away with `guards` or `continue`  like you've done above.


```suggestion
				reqd, n := requiredBySuccessor(attr, node, succ.(*PhysicalPlanNode));
				
				if reqd {
				   continue;
				}
				
 				if n != nil {
				   return &flux.Error{
					Code: codes.Internal,
					Msg: fmt.Sprintf("plan node %q has attribute %q that must be required by successors, "+
				        "but it is not required or propagated by successor %q",
				        node.ID(), attr.Key(), n.ID(),
				   ),
			        }
				return &flux.Error{
					Code: codes.Internal,
					Msg: fmt.Sprintf("plan node %q has attribute %q that must be required by successors, "+
						"but no successors require it",
						node.ID(), attr.Key(),
					),
				}	
```

Reply:
Good idea, I've made the change.